### PR TITLE
runs informer integration tests in CI

### PR DESCRIPTION
**Pull request by xh3b4sd**
Towards https://github.com/giantswarm/giantswarm/issues/2774.

### CODE REVIEW

**Comment by xh3b4sd** [comment]
@fgimenez I got an issue here. See https://circleci.com/gh/giantswarm/operatorkit/488. 

```
$ #!/bin/bash -eo pipefail
./e2e-harness test --test-dir=informer/integration
{"caller":"github.com/giantswarm/e2e-harness/pkg/compiler/compiler.go:64","info":"Compiling binary operatorkit-e2e","time":"2018-03-09 12:33:43.397"}
{"caller":"github.com/giantswarm/e2e-harness/pkg/compiler/compiler.go:45","info":"Compiling binary operatorkit","time":"2018-03-09 12:34:23.953"}
{"caller":"github.com/giantswarm/e2e-harness/pkg/compiler/compiler.go:47","info":"error compiling binary operatorkit","time":"2018-03-09 12:34:23.987"}
Error: exit status 1
Usage:
  e2e-harness test [flags]

Flags:
  -h, --help              help for test
      --test-dir string   Name of the directory containing executable tests. (default "integration")

exit status 1
Exited with code 1
```

Can you shed some light on why this is happening? I am not sure how to deal with this. I would even say there is no need to build operatorkit as a binary.

**Comment by fgimenez** [comment]
Exactly, in this case it shouldn't try to build the binary. Looks like we need a `--skip-build` flag, I'll prepare it. For the time being you can workaround this issue by creating an `operatorkit` executable file before calling `e2e-harness test`, something like `touch ./operatorkit && chmod a+x ./operatorkit`.

**Comment by xh3b4sd** [comment]
I will try the workaround for now.

**Comment by xh3b4sd** [comment]
It now fails building the image. See https://circleci.com/gh/giantswarm/operatorkit/490. 

```
$ #!/bin/bash -eo pipefail
./e2e-harness test --test-dir=informer/integration
{"caller":"github.com/giantswarm/e2e-harness/pkg/compiler/compiler.go:64","info":"Compiling binary operatorkit-e2e","time":"2018-03-09 13:32:37.496"}
{"caller":"github.com/giantswarm/e2e-harness/pkg/compiler/compiler.go:45","info":"Compiling binary operatorkit","time":"2018-03-09 13:33:16.095"}
{"caller":"github.com/giantswarm/e2e-harness/pkg/compiler/compiler.go:80","function":"compileMain","level":"info","message":"main binary exists, not building","time":"2018-03-09 13:33:16.095"}
{"caller":"github.com/giantswarm/e2e-harness/pkg/minikube/minikube.go:35","info":"Getting minikube docker environment","time":"2018-03-09 13:33:16.095"}
{"caller":"github.com/giantswarm/e2e-harness/pkg/minikube/minikube.go:45","info":"Building image quay.io/giantswarm/operatorkit","time":"2018-03-09 13:33:16.096"}
error building image quay.io/giantswarm/operatorkit
Error: exit status 1
Usage:
  e2e-harness test [flags]

Flags:
  -h, --help              help for test
      --test-dir string   Name of the directory containing executable tests. (default "integration")

exit status 1
Exited with code 1
```

**Comment by fgimenez** [comment]
This should fix the build issues https://github.com/giantswarm/e2e-harness/pull/27

**Comment by xh3b4sd** [comment]
Super cool. Seems like it worked. The output is just not really helpful. I think we should display the test and all the `go test` output. I remember other tests having this. Why is this dropped here? 

```
{"caller":"github.com/giantswarm/e2e-harness/pkg/project/project.go:139","info":"started tests","time":"2018-03-09 14:05:39.846"}
PASS
{"caller":"github.com/giantswarm/e2e-harness/pkg/project/project.go:156","info":"finished tests","time":"2018-03-09 14:06:03.135"}
```

File: .circleci/config.yml [hunk 1]
```diff
@@ -15,3 +15,30 @@ jobs:
           if [ "${CIRCLE_BRANCH}" == "master" ]; then
             ./architect deploy
           fi
+
+  e2eTestInformerIntegration:
```
Comment (xh3b4sd) [inline .circleci/config.yml]: We register a new integration test here.

File: .circleci/config.yml [hunk 2]
```diff
@@ -15,3 +15,30 @@ jobs:
...
+    machine: true
+    working_directory: /home/circleci/.go_workspace/src/github.com/giantswarm/operatorkit
+    environment:
+      MINIKUBE_VERSION: v0.25.0
```
Comment (xh3b4sd) [inline .circleci/config.yml]: I feel like this should be a default in e2e-harness. We can then switch to newer versions everywhere more easily.

File: .circleci/config.yml [hunk 3]
```diff
@@ -15,3 +15,30 @@ jobs:
...
+    steps:
+    - checkout
+
+    - run: |
+        wget -q $(curl -sS https://api.github.com/repos/giantswarm/e2e-harness/releases/latest | grep browser_download_url | head -n 1 | cut -d '"' -f 4)
+        chmod +x ./e2e-harness
+    - run: ./e2e-harness localkube
+    - run: ./e2e-harness setup --remote=false
+    - run: ./e2e-harness test --test-dir=informer/integration
```
Comment (xh3b4sd) [inline .circleci/config.yml]: Here we configure the specific folder.

File: Gopkg.toml [hunk 4]
```diff
@@ -24,6 +24,10 @@
   name = "github.com/cenkalti/backoff"
   version = "1.1.0"
 
+[[constraint]]
+  branch = "master"
+  name = "github.com/giantswarm/e2e-harness"
```
Comment (xh3b4sd) [inline Gopkg.toml]: New dep.

File: informer/informer_integration_test.go [hunk 5]
```diff
@@ -1,233 +0,0 @@
...
-
-		-integration.address string
-			Kubernetes API server address (default "https://$(minikube ip):8443")
-		-integration.ca string
-			CA file path (default "$HOME/.minikube/ca.crt")
-		-integration.crt string
-			certificate file path (default "$HOME/.minikube/apiserver.crt")
-		-integration.key string
-			key file path (default "$HOME/.minikube/apiserver.key")
-*/
```
Comment (xh3b4sd) [inline informer/informer_integration_test.go]: So far we executed against a local minikube. This is now in CI and I did not find a cool and fast way to keep support for both. Ideas welcome.

File: informer/integration/basic_test.go [hunk 6]
```diff
@@ -12,24 +12,29 @@ import (
...
 func Test_Informer_Integration_Basic(t *testing.T) {
-	testSetup(t)
-	defer testTeardown(t)
-
-	const timeDelta = time.Millisecond * 100
-
-	objectIDOne := "al7qy"
-	objectIDTwo := "al8qy"
+	mustSetup()
+	defer mustTeardown()
```
Comment (xh3b4sd) [inline informer/integration/basic_test.go]: I also did some refactoring on the test itself. It tests the same things. Just some code changes to make me happy.

File: informer/integration/project.yaml [hunk 7]
```diff
@@ -0,0 +1,4 @@
+version: 1
+
+test:
+  env:
```
Comment (xh3b4sd) [inline informer/integration/project.yaml]: We do not need this here I think. Can we just drop it though? Does e2e-harness support this?

**Comment by fgimenez** [comment]
> I think we should display the test and all the go test output. I remember other tests having this. Why is this dropped here?

We are calling the test binary without any additional flag for all the projects https://github.com/giantswarm/e2e-harness/blob/master/pkg/project/project.go#L152 some of them add logging inside the tests. I agree it would be good to pass the verbose flag for all of them.

File: informer/informer_integration_test.go (hunk 5 above)
Comment (kopiczko) [reply informer/informer_integration_test.go]: I think it's possible to run that with e2e-tests binary. @fgimenez ?

**Comment by xh3b4sd** [comment]
> I agree it would be good to pass the verbose flag for all of them.

Then lets just do this. Having this information helps reasoning about integration tests have run properly. In this case here I would like to see the test takes about 25 seconds. It might end faster by accident and is broken but we do not know now.

File: informer/informer_integration_test.go (hunk 5 above)
Comment (fgimenez) [reply informer/informer_integration_test.go]: In fact e2e-harness allows you to run on a local minikube too without changing anything else, having e2e-harness binary just `minikube delete && minikube start --extra-config=apiserver.Authorization.Mode=RBAC && e2e-harness setup --remote=false && e2e-harness test`. This is why it tries to build the project binary and the image, and also why the tests are run inside a container, it works both locally and in CI seamlessly.

File: informer/integration/project.yaml (hunk 7 above)
Comment (fgimenez) [reply informer/integration/project.yaml]: Currently it expects to find the `integration/project.yaml` file https://github.com/giantswarm/e2e-harness/blob/master/pkg/project/project.go#L187, the `env` section can be safely dropped https://github.com/giantswarm/e2e-harness/blob/master/pkg/project/project.go#L200

File: informer/informer_integration_test.go (hunk 5 above)
Comment (xh3b4sd) [reply informer/informer_integration_test.go]: Nice.

File: informer/integration/project.yaml (hunk 7 above)
Comment (xh3b4sd) [reply informer/integration/project.yaml]: Should we make this optional then as well?

File: informer/integration/custom_resource.go [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+	}
+
+	return &CustomResourceList{
+		TypeMeta: t.TypeMeta,
+		ListMeta: *t.ListMeta.DeepCopy(),
+		Items:    itemsCopy,
+	}
+}
+
+func createCustomResource(ID string) error {
```
Comment (kopiczko) [inline informer/integration/custom_resource.go]: It doesn't create a custom resource. It's a ConfigMap.

File: informer/integration/custom_resource.go [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+
+const (
+	namespace = "informer-integration-test"
+)
+
+type CustomResource struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+	Spec              CustomResourceSpec `json:"spec"`
+}
```
Comment (kopiczko) [inline informer/integration/custom_resource.go]: I don't think this type is needed, probably some copied dead code.

File: informer/integration/custom_resource.go [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+	}
+
+	return nil
+}
+
+func mustAssertCRWithID(e watch.Event, IDs ...string) {
+	m, err := meta.Accessor(e.Object)
+	if err != nil {
+		panic(err)
+	}
```
Comment (kopiczko) [inline informer/integration/custom_resource.go]: Object can be case to ConfigMap to be more explicit.

File: informer/integration/project.yaml (hunk 7 above)
Comment (fgimenez) [reply informer/integration/project.yaml]: :+1: I'll prepare the change.

File: informer/integration/basic_test.go [hunk 11]
```diff
@@ -12,24 +12,29 @@ import (
...
 
+	crIDOne := "al7qy"
+	crIDTwo := "al8qy"
```
Comment (kopiczko) [inline informer/integration/basic_test.go]: The object is configmap, so I think cmID is more accurate prefix.

**Comment by xh3b4sd** [comment]
I noticed this but not sure what this is about. What can we do about it?

**Comment by fgimenez** [comment]
> I noticed this but not sure what this is about. What can we do about it?

It needs to be configured in Settings -> Branches -> master Edit -> select the jobs required (remove ci/circleci there and optionally set the others as required).

Also related to this, once we are happy with how the tests work we might make the deployment only when the tests have passed.

**Comment by xh3b4sd** [comment]
Oh I see this is just Github magic. Why is this now a problem though? I do not get that.

**Comment by xh3b4sd** [comment]
Oh because the circle config changed. With you.

**Comment by xh3b4sd** [comment]
BAM. I like this.

File: informer/integration/custom_resource.go (hunk 8 above)
Comment (xh3b4sd) [reply informer/integration/custom_resource.go]: I remember you improved this to use the config maps at some point. Cleaned up now.

File: informer/integration/custom_resource.go (hunk 9 above)
Comment (xh3b4sd) [reply informer/integration/custom_resource.go]: Gone.

File: informer/integration/custom_resource.go (hunk 10 above)
Comment (xh3b4sd) [reply informer/integration/custom_resource.go]: Done.

File: informer/integration/basic_test.go (hunk 11 above)
Comment (xh3b4sd) [reply informer/integration/basic_test.go]: Done.

File: .circleci/config.yml (hunk 2 above)
Comment (xh3b4sd) [reply .circleci/config.yml]: @fgimenez do we have this somewhere tracked?

File: informer/integration/custom_resource.go (hunk 8 above)
Comment (kopiczko) [reply informer/integration/custom_resource.go]: Yes, I most likely forgot to delete the old struct, and rename few other things ;)

File: .gitignore [hunk 12]
```diff
@@ -0,0 +1,5 @@
+TODO
+!vendor/**
+
+.e2e-harness/
+informer/integration/operatorkit-e2e
```
Comment (xh3b4sd) [inline .gitignore]: I added this.

File: .circleci/config.yml (hunk 2 above)
Comment (fgimenez) [reply .circleci/config.yml]: Just added it here https://github.com/giantswarm/e2e-harness/issues/29

**Comment by xh3b4sd** [comment]
I go with this now. This is cool.

**Pull request merged**